### ISSUE TITLE
Git tagging failing with: Not a git repository (or any parent up to mount point /workspace)

### ISSUE THREAD

**Issue by dbolkensteyn** [GoogleContainerTools/skaffold#2067]
### Expected behavior
Images get tagged with the commit SHA1

### Actual behavior
Images get tagged as `dirty`.

### Information

I am triggering Skaffold from Google Cloud Build through the GitHub app, using Google Cloud Build to build the various Docker images in parallel.

`cloudbuild.yaml`
```yaml
steps:
- name: 'gcr.io/k8s-skaffold/skaffold'
  args: [ 'skaffold', 'build', '-p', 'cloud-build']
```

`skaffold.yaml`
```yaml
apiVersion: skaffold/v1beta8
kind: Config
profiles:
- name: cloud-build
  build:
    googleCloudBuild:
      machineType: N1_HIGHCPU_32
build:
  artifacts:
  - image: eu.gcr.io/project_id/service1
    context: src/service1/
  - image: eu.gcr.io/project_id/service2
    context: src/service2/
```

If I download the .tar.gz sources for the Google Cloud Build dedicated to `eu.gcr.io/project_id/service1`, I only find the contents under `src/service1`. How is Skaffold supposed to find the Git commit back here?

### Steps to reproduce the behavior

1. Run such a build
2. `eu.gcr.io/project_id/service1` and `eu.gcr.io/project_id/service2` both will get tagged as dirty, whereas I was expecting to be able to trace the commit's SHA1

**Comment by dbolkensteyn** [comment]
So the build step received the following arguments:

```
**Name** gcr.io/cloud-builders/docker
**Arguments** build --tag eu.gcr.io/project_id/service1:dirty -f Dockerfile .
```

So Skaffold somehow already determined that the tag was dirty from the main dispatching Cloud Build.

**Comment by dbolkensteyn** [comment]
And there are a bunch of such warnings in the logs of the main dispatching Cloud Build run:
```
time="2019-05-03T22:26:44Z" level=warning msg="Unable to find git commit: Running [git describe --tags --always]: stdout , stderr: fatal: Not a git repository (or any parent up to mount point /workspace)\nStopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).\n, err: exit status 128: exit status 128"
```

**Comment by dbolkensteyn** [comment]
So I am seeing the error about with both `latest` and `v0.28.0` tags of the `gcr.io/k8s-skaffold/skaffold` image.

Now, if I downgrade to `v0.27.0`, I am hitting error #1979 instead (`exec: \"git\": executable file not found in $PATH`).

I'll try to find a SHA1 for the image that includes https://github.com/GoogleContainerTools/skaffold/pull/1987, prior to the migration to go-git to try to see if that is causes this problem.

**Comment by dbolkensteyn** [comment]
Meanwhile, as a workaround, one can rely on Google Cloud Build's `$REVISION_ID` environment variable in the `skaffold.yaml` to force the tag:

```yaml
profiles:
- name: cloud-build
  build:
    googleCloudBuild:
      machineType: N1_HIGHCPU_32
    tagPolicy:
      envTemplate:
        template: "{{.IMAGE_NAME}}:{{.REVISION_ID}}"
...
```

and by forwarding the `$REVISION_ID` environment variable from the `cloudbuild.yaml` file:
```yaml
steps:
- name: 'gcr.io/k8s-skaffold/skaffold'
  args: [ 'skaffold', 'build', '-p', 'cloud-build' ]
  env:
  - 'REVISION_ID=$REVISION_ID'
```

**Comment by dbolkensteyn** [comment]
Note that this might actually be a problem with the Google Cloud Build GitHub app itself instead of Skaffold, which seem to fetch the repo sources from downloading a tarball rather than performing a Git clone, and hence not recreate a properly formed Git repository.

**Comment by tejal29** [comment]
The way skaffold with GCB works is
1. From the config, skaffold will create GCB builder
2. From the config, it will create the Tagger to use (in your case Git Tagger)
3. Skaffold will call GCBBuilder.Buildr(ctx, out, tags, artifactsToBuild)
   where
   1. ctx is the context
   2. out is where output is written
   3. tags: is calculated images tags as per the Tagging  Policy.
In short, tags are computed locally and sent to the GCB builder. 

Can you run your command skaffold locally instead of using the google cloud builder and see if the tag is correct?

Also, regarding your use case
>I am triggering Skaffold from Google Cloud Build through the GitHub app, using Google Cloud Build to build the various Docker images in parallel.

Skaffold google cloud builder already run builds in parallel. 
https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/build/gcb/cloud_build.go#L47

**Comment by balopat** [comment]
1. I'd like to understand if the Github App is really the culprit in getting the repo without the .git directory - or if there is a filtering in place somewhere maybe? Do you have any.gitignore, .dockerignore, .gcloudignore files? 
2. Currently you have GCB that kicks off another GCB process through skaffold - you could just use use a simple docker builder in skaffold (without the `-p cloud-build` profile). Did you try that?

**Comment by balopat** [comment]
I'm closing this for now, please reopen it if you think there is something else we can improve.

**Comment by dbolkensteyn** [comment]
Sorry for the late follow-up @tejal29 @balopat 

There seems to be some confusion because:
1. Skaffold can call Cloud Build: From my local machine, I want to offload builds to the cloud
2. Cloud Build can call Skaffold: I want a CI/CD pipeline on every push to GitHub

From my local machine, running Skaffold with a `googleCloudBuild` build works as expected and images get tagged with the right Git commit.

The issue occurs in the second scenario. The full pipeline actually looks like this:
Push to GitHub -> Google Cloud Build (invoked through a webhook by GitHub) -> Skaffold (controlled by `cloudbuild.yaml`) -> Google Cloud Build (triggered by Skaffold this time, to parallelize the build, controlled by `skaffold.yml`)

The problem is with the first Google Cloud Build, the one triggered by GitHub: It doesn't clone/fetch the sources from GitHub using Git itself, but downloads a tarball from a Bucket instead. The trigger name in this scenario is set to "Default GitHub Checks". The very first line in the build logs is:
```
Fetching storage object: gs://[digits].cloudbuild-source.googleusercontent.com/[commit SHA]-[GUID].tar.gz#[digits]
```

Then, when `skaffold run` runs a couple line later, it of course fails to locate the `.git` repository and outputs:
```
Step #1: time="2019-05-29T15:47:02Z" level=warning msg="Unable to find git commit: Running [git describe --tags --always]: stdout , stderr: fatal: Not a git repository (or any parent up to mount point /workspace)\nStopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).\n, err: exit status 128: exit status 128"
```

Long story short, I believe that this issue is more a Google Cloud Build one than a Skaffold one, and I see 2 possible options to fix it:
1. Google Cloud Build clones the repository from GitHub using Git instead of downloading a tarball
2. A dedicated tagger is added for Google Cloud Build to make use of the `$REVISION_ID` environment variable that it sets

Does this make more sense now?

**Comment by tejal29** [comment]
You can make use of [profiles](https://skaffold.dev/docs/how-tos/profiles/) to use [environment tagger](https://skaffold.dev/docs/how-tos/taggers/#envtemplate-uses-values-of-environment-variables-as-tags). for Step 3. 
Step 3 will now be `skaffold build -p env_tagger_gcb_build`


Google cloud build also specifies a way to clone your git repo and it can be done via [cloud build  step](https://cloud.google.com/cloud-build/docs/build-config#build_steps).
```
steps:
- name: 'gcr.io/cloud-builders/git'
  args: ['clone', 'https://github.com/GoogleCloudPlatform/cloud-builders']
  env: ['PROJECT_ROOT=hello']
- name: 'gcr.io/cloud-builders/docker'
  args: ['build', '-t', 'gcr.io/my-project-id/myimage', '.']
```

**Comment by tejal29** [comment]
on wait, you are already doing what i mentioned as the first solution.

**Comment by dbolkensteyn** [comment]
Thank you @tejal29 , I was not aware of the possibility to use an explicit git build step.

Consider this solved 😄
